File: tests/AetherNet.Core.Tests/DtnFixtureTests.cs
// SPDX-License-Identifier: MIT

using System.Text.Json;
using System.Text.Json.Serialization;
using AetherNet.Dtn;
using AetherNet.Models;
using Xunit;

namespace AetherNet.Core.Tests;

/// <summary>
/// Cross-language DTN-envelope wire-format verifier. Reads fixtures/dtn/inputs.json
/// and fixtures/dtn/expected/*.bin (the Go-oracle output, committed to the repo) and
/// asserts that this language's <see cref="DtnEnvelopeSerializer"/> produces the same
/// bytes for each canonical input and round-trips every field.
/// </summary>
public class DtnFixtureTests
{
    private record DtnFixtureInput(
        [property: JsonPropertyName("kind")] string Kind,
        [property: JsonPropertyName("name")] string Name,
        [property: JsonPropertyName("id")] string? Id,
        [property: JsonPropertyName("priority")] int Priority,
        [property: JsonPropertyName("status")] int Status,
        [property: JsonPropertyName("copy_count")] int CopyCount,
        [property: JsonPropertyName("max_copies")] int MaxCopies,
        [property: JsonPropertyName("hop_count")] int HopCount,
        [property: JsonPropertyName("created_at_ms")] long CreatedAtMs,
        [property: JsonPropertyName("expires_at_ms")] long ExpiresAtMs,
        [property: JsonPropertyName("sender_uhid")] string? SenderUhid,
        [property: JsonPropertyName("recipient_uhid")] string? RecipientUhid,
        [property: JsonPropertyName("sender_geohash")] string? SenderGeohash,
        [property: JsonPropertyName("recipient_last_geohash")] string? RecipientLastGeohash,
        [property: JsonPropertyName("encrypted_payload_hex")] string? EncryptedPayloadHex,
        [property: JsonPropertyName("encrypted_payload_len")] int EncryptedPayloadLen,
        [property: JsonPropertyName("bundle_id")] string? BundleId,
        [property: JsonPropertyName("accepted")] bool Accepted,
        [property: JsonPropertyName("total_hops")] int TotalHops,
        [property: JsonPropertyName("total_custody_transfers")] int TotalCustodyTransfers,
        [property: JsonPropertyName("delivered_at_ms")] long DeliveredAtMs);

    private static byte[] HexToBytes(string? hex)
    {
        if (string.IsNullOrEmpty(hex)) return Array.Empty<byte>();
        var n = hex.Length / 2;
        var bytes = new byte[n];
        for (var i = 0; i < n; i++)
            bytes[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
        return bytes;
    }

    private static byte[] PayloadFor(DtnFixtureInput input)
    {
        if (input.EncryptedPayloadLen > 0)
        {
            var b = new byte[input.EncryptedPayloadLen];
            for (var i = 0; i < b.Length; i++) b[i] = (byte)(i % 256);
            return b;
        }
        return HexToBytes(input.EncryptedPayloadHex);
    }

    private static string FixturesDir()
    {
        var dir = AppContext.BaseDirectory;
        for (var i = 0; i < 10; i++)
        {
            var candidate = Path.Combine(dir, "fixtures", "dtn", "inputs.json");
            if (File.Exists(candidate)) return Path.Combine(dir, "fixtures", "dtn");
            var parent = Path.GetDirectoryName(dir.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar));
            if (parent is null || parent == dir) break;
            dir = parent;
        }
        throw new FileNotFoundException("Could not locate fixtures/dtn/inputs.json from " + AppContext.BaseDirectory);
    }

    private static List<DtnFixtureInput> LoadInputs()
    {
        var path = Path.Combine(FixturesDir(), "inputs.json");
        return JsonSerializer.Deserialize<List<DtnFixtureInput>>(File.ReadAllText(path))!;
    }

    public static IEnumerable<object[]> AllFixtures() =>
        LoadInputs().Select(x => new object[] { x.Name });

    private static byte[] Serialize(DtnFixtureInput input) => input.Kind switch
    {
        "bundle" => DtnEnvelopeSerializer.SerializeBundle(new DtnBundle
        {
            Id = Guid.Parse(input.Id!),
            SenderUhid = input.SenderUhid ?? string.Empty,
            RecipientUhid = input.RecipientUhid ?? string.Empty,
            EncryptedPayload = PayloadFor(input),
            Priority = (BundlePriority)(byte)input.Priority,
            Status = (BundleStatus)(byte)input.Status,
            CopyCount = input.CopyCount,
            MaxCopies = input.MaxCopies,
            SenderGeohash = input.SenderGeohash,
            RecipientLastGeohash = input.RecipientLastGeohash,
            HopCount = input.HopCount,
            CreatedAt = DateTimeOffset.FromUnixTimeMilliseconds(input.CreatedAtMs).UtcDateTime,
            ExpiresAt = DateTimeOffset.FromUnixTimeMilliseconds(input.ExpiresAtMs).UtcDateTime,
        }),
        "custody_ack" => DtnEnvelopeSerializer.SerializeCustodyAck(Guid.Parse(input.BundleId!), input.Accepted),
        "delivery_receipt" => DtnEnvelopeSerializer.SerializeDeliveryReceipt(new DtnDeliveryReceipt
        {
            BundleId = Guid.Parse(input.BundleId!),
            RecipientUhid = input.RecipientUhid ?? string.Empty,
            TotalHops = input.TotalHops,
            TotalCustodyTransfers = input.TotalCustodyTransfers,
            DeliveredAt = DateTimeOffset.FromUnixTimeMilliseconds(input.DeliveredAtMs).UtcDateTime,
        }),
        _ => throw new InvalidOperationException($"unknown kind {input.Kind}"),
    };

    [Theory]
    [MemberData(nameof(AllFixtures))]
    public void Serialize_MatchesExpectedBytes(string name)
    {
        var input = LoadInputs().Single(x => x.Name == name);
        var serialized = Serialize(input);
        var expected = File.ReadAllBytes(Path.Combine(FixturesDir(), "expected", name + ".bin"));
        Assert.Equal(expected, serialized);
    }

    [Theory]
    [MemberData(nameof(AllFixtures))]
    public void Deserialize_FromExpectedBytes_MatchesInputFields(string name)
    {
        var input = LoadInputs().Single(x => x.Name == name);
        var bytes = File.ReadAllBytes(Path.Combine(FixturesDir(), "expected", name + ".bin"));

        switch (input.Kind)
        {
            case "bundle":
                var b = DtnEnvelopeSerializer.DeserializeBundle(bytes);
                Assert.Equal(Guid.Parse(input.Id!), b.Id);
                Assert.Equal((BundlePriority)(byte)input.Priority, b.Priority);
                Assert.Equal((BundleStatus)(byte)input.Status, b.Status);
                Assert.Equal(input.CopyCount, b.CopyCount);
                Assert.Equal(input.MaxCopies, b.MaxCopies);
                Assert.Equal(input.HopCount, b.HopCount);
                Assert.Equal(input.CreatedAtMs, new DateTimeOffset(b.CreatedAt, TimeSpan.Zero).ToUnixTimeMilliseconds());
                Assert.Equal(input.ExpiresAtMs, new DateTimeOffset(b.ExpiresAt, TimeSpan.Zero).ToUnixTimeMilliseconds());
                Assert.Equal(input.SenderUhid ?? string.Empty, b.SenderUhid);
                Assert.Equal(input.RecipientUhid ?? string.Empty, b.RecipientUhid);
                Assert.Equal(input.SenderGeohash ?? string.Empty, b.SenderGeohash);
                Assert.Equal(input.RecipientLastGeohash ?? string.Empty, b.RecipientLastGeohash);
                Assert.Equal(PayloadFor(input), b.EncryptedPayload);
                break;
            case "custody_ack":
                var (ackId, accepted) = DtnEnvelopeSerializer.DeserializeCustodyAck(bytes);
                Assert.Equal(Guid.Parse(input.BundleId!), ackId);
                Assert.Equal(input.Accepted, accepted);
                break;
            case "delivery_receipt":
                var r = DtnEnvelopeSerializer.DeserializeDeliveryReceipt(bytes);
                Assert.Equal(Guid.Parse(input.BundleId!), r.BundleId);
                Assert.Equal(input.RecipientUhid ?? string.Empty, r.RecipientUhid);
                Assert.Equal(input.TotalHops, r.TotalHops);
                Assert.Equal(input.TotalCustodyTransfers, r.TotalCustodyTransfers);
                Assert.Equal(input.DeliveredAtMs, new DateTimeOffset(r.DeliveredAt, TimeSpan.Zero).ToUnixTimeMilliseconds());
                break;
        }
    }
}

File: rust/tests/dtn_fixture_tests.rs
// SPDX-License-Identifier: MIT
//! Cross-language DTN-envelope fixture verifier.
//!
//! Reads `../fixtures/dtn/inputs.json` + `../fixtures/dtn/expected/*.bin` and
//! asserts this crate's binary DTN envelope is byte-identical to the Go oracle
//! (`go/cmd/dtnfixturegen`) for every case, then round-trips each back to
//! matching fields. See `fixtures/README.md`.

use std::fs;
use std::path::{Path, PathBuf};

use serde_json::Value;
use uuid::Uuid;

use aethernet_protocol::dtn::envelope::{
    deserialize_bundle, deserialize_custody_ack, deserialize_delivery_receipt, serialize_bundle,
    serialize_custody_ack, serialize_delivery_receipt,
};
use aethernet_protocol::models::{BundlePriority, BundleStatus, DtnBundle};

fn fixtures_dir() -> PathBuf {
    // CARGO_MANIFEST_DIR = .../rust/, parent = aether-protocol/
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .join("fixtures")
        .join("dtn")
}

fn load_inputs() -> Vec<Value> {
    let path = fixtures_dir().join("inputs.json");
    let raw = fs::read_to_string(&path).unwrap_or_else(|e| panic!("read {:?}: {}", path, e));
    serde_json::from_str(&raw).expect("parse inputs.json")
}

fn expected(name: &str) -> Vec<u8> {
    let p = fixtures_dir().join("expected").join(format!("{}.bin", name));
    fs::read(&p).unwrap_or_else(|e| panic!("read {:?}: {}", p, e))
}

fn hex_decode(s: &str) -> Vec<u8> {
    (0..s.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap())
        .collect()
}

fn payload_for(o: &Value) -> Vec<u8> {
    let len = o.get("encrypted_payload_len").and_then(Value::as_u64).unwrap_or(0) as usize;
    if len > 0 {
        return (0..len).map(|i| (i % 256) as u8).collect();
    }
    hex_decode(o.get("encrypted_payload_hex").and_then(Value::as_str).unwrap_or(""))
}

fn str_field<'a>(o: &'a Value, key: &str) -> &'a str {
    o.get(key).and_then(Value::as_str).unwrap_or("")
}

fn i32_field(o: &Value, key: &str) -> i32 {
    o.get(key).and_then(Value::as_i64).unwrap_or(0) as i32
}

fn i64_field(o: &Value, key: &str) -> i64 {
    o.get(key).and_then(Value::as_i64).unwrap_or(0)
}

fn nullable_str(o: &Value, key: &str) -> Option<String> {
    match o.get(key) {
        Some(Value::String(s)) if !s.is_empty() => Some(s.clone()),
        _ => None,
    }
}

fn uuid_field(o: &Value, key: &str) -> Uuid {
    Uuid::parse_str(str_field(o, key)).unwrap()
}

fn bundle_from(o: &Value) -> DtnBundle {
    DtnBundle {
        id: uuid_field(o, "id"),
        sender_uhid: str_field(o, "sender_uhid").to_string(),
        recipient_uhid: str_field(o, "recipient_uhid").to_string(),
        encrypted_payload: payload_for(o),
        priority: BundlePriority::from_u8(i32_field(o, "priority") as u8),
        status: BundleStatus::from_u8(i32_field(o, "status") as u8),
        copy_count: i32_field(o, "copy_count"),
        max_copies: i32_field(o, "max_copies"),
        sender_geohash: nullable_str(o, "sender_geohash"),
        recipient_last_geohash: nullable_str(o, "recipient_last_geohash"),
        hop_count: i32_field(o, "hop_count"),
        created_at: (i64_field(o, "created_at_ms") / 1000) as u64,
        expires_at: (i64_field(o, "expires_at_ms") / 1000) as u64,
    }
}

fn kind(o: &Value) -> &str {
    o.get("kind").and_then(Value::as_str).unwrap()
}

fn serialize_case(o: &Value) -> Vec<u8> {
    match kind(o) {
        "bundle" => serialize_bundle(&bundle_from(o)),
        "custody_ack" => serialize_custody_ack(
            &uuid_field(o, "bundle_id"),
            o.get("accepted").and_then(Value::as_bool).unwrap_or(false),
        ),
        "delivery_receipt" => serialize_delivery_receipt(
            &uuid_field(o, "bundle_id"),
            str_field(o, "recipient_uhid"),
            i32_field(o, "total_hops"),
            i32_field(o, "total_custody_transfers"),
            i64_field(o, "delivered_at_ms"),
        ),
        other => panic!("unknown kind {}", other),
    }
}

#[test]
fn dtn_serialize_matches_go_oracle() {
    for o in load_inputs() {
        let name = str_field(&o, "name");
        assert_eq!(
            serialize_case(&o),
            expected(name),
            "fixture {}: serialize bytes diverge — see fixtures/README.md",
            name
        );
    }
}

#[test]
fn dtn_deserialize_roundtrips_all_fields() {
    for o in load_inputs() {
        let name = str_field(&o, "name");
        let data = expected(name);
        match kind(&o) {
            "bundle" => {
                let b = deserialize_bundle(&data)
                    .unwrap_or_else(|| panic!("{}: deserialize_bundle returned None", name));
                assert_eq!(b.id, uuid_field(&o, "id"), "{}: id", name);
                assert_eq!(b.priority.as_u8(), i32_field(&o, "priority") as u8, "{}: priority", name);
                assert_eq!(b.status.as_u8(), i32_field(&o, "status") as u8, "{}: status", name);
                assert_eq!(b.copy_count, i32_field(&o, "copy_count"), "{}: copy_count", name);
                assert_eq!(b.max_copies, i32_field(&o, "max_copies"), "{}: max_copies", name);
                assert_eq!(b.hop_count, i32_field(&o, "hop_count"), "{}: hop_count", name);
                assert_eq!(
                    b.created_at,
                    (i64_field(&o, "created_at_ms") / 1000) as u64,
                    "{}: created_at",
                    name
                );
                assert_eq!(
                    b.expires_at,
                    (i64_field(&o, "expires_at_ms") / 1000) as u64,
                    "{}: expires_at",
                    name
                );
                assert_eq!(b.sender_uhid, str_field(&o, "sender_uhid"), "{}: sender_uhid", name);
                assert_eq!(b.recipient_uhid, str_field(&o, "recipient_uhid"), "{}: recipient_uhid", name);
                assert_eq!(
                    b.sender_geohash.as_deref().unwrap_or(""),
                    str_field(&o, "sender_geohash"),
                    "{}: sender_geohash",
                    name
                );
                assert_eq!(
                    b.recipient_last_geohash.as_deref().unwrap_or(""),
                    str_field(&o, "recipient_last_geohash"),
                    "{}: recipient_last_geohash",
                    name
                );
                assert_eq!(b.encrypted_payload, payload_for(&o), "{}: payload", name);
            }
            "custody_ack" => {
                let (id, accepted) =
                    deserialize_custody_ack(&data).unwrap_or_else(|| panic!("{}: custody_ack", name));
                assert_eq!(id, uuid_field(&o, "bundle_id"), "{}: bundle_id", name);
                assert_eq!(
                    accepted,
                    o.get("accepted").and_then(Value::as_bool).unwrap_or(false),
                    "{}: accepted",
                    name
                );
            }
            "delivery_receipt" => {
                let (id, recipient, hops, transfers, delivered) =
                    deserialize_delivery_receipt(&data).unwrap_or_else(|| panic!("{}: receipt", name));
                assert_eq!(id, uuid_field(&o, "bundle_id"), "{}: bundle_id", name);
                assert_eq!(recipient, str_field(&o, "recipient_uhid"), "{}: recipient_uhid", name);
                assert_eq!(hops, i32_field(&o, "total_hops"), "{}: total_hops", name);
                assert_eq!(
                    transfers,
                    i32_field(&o, "total_custody_transfers"),
                    "{}: total_custody_transfers",
                    name
                );
                assert_eq!(delivered, i64_field(&o, "delivered_at_ms"), "{}: delivered_at_ms", name);
            }
            other => panic!("unknown kind {}", other),
        }
    }
}
